### DISCUSSION TITLE
Conditional `use`

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#136] [Q&A]
Would you advice against this conditional `use` knowing that the condition is assured to be always the same (immutable)? Is the second time I do like this. I know I could "use" it in a not conditional way, but then the object will be created even when it's not needed.

```dart
Router _rootRouterCapsule(CapsuleHandle use) {
  final config = use(appConfigCapsule);

  final router = Router(
    notFoundHandler: (_) => Response.badRequest(body: 'Bad root API request'),
  );

  if (config.cloudflareR2Enabled) {
    router.mount('/cloud-assets', use(cloudAssetsRouterCapsule).call);
  }

  return router;
}
```

**Comment by GregoryConrad** [comment]
I don’t see anything wrong there.

It’s actually a best practice to conditionally use(someOtherCapsule) when you only need it some of the time.

the same is not true for use.foobar (side effects)—they must always be called unconditionally due to the implementation.

**Reply by busslina** [reply]
Cool, I mixed `CapsuleReader` with `SideEffectRegistrar`. So now it came to me two new questions:

- If the "used" capsule rebuilds, will the "using" capsule rebuild too only if the "using" condition was true on the last build?
- Does ReArch detect circular dependencies (or loops)? I suppose it has to fail on that case.

**Reply by GregoryConrad** [reply]
1. Correct. A capsule’s dependencies are updated each time it builds, so if it doesn’t use a capsule on one particular build, that capsule will not cause it to rebuild.
2. There isn’t a hard check for circular dependencies at the moment but you’ll get a stack overflow or something.
